api/src/services/versions.ts

```diff
@@ -123,6 +123,10 @@ export class VersionsService extends ItemsService {
 
 		if (!versions?.[0]) return null;
 
+		if (versions[0]['delta']) {
+			return [versions[0]['delta']];
+		}
+
 		const saves = await this.getVersionSavesById(versions[0]['id']);
```

Comment:
This `getVersionSavesById ` is just for the edge case where a new version is created but on the older system when the migration is being run. All usages of it are part of the else clause where the `delta` is missing. 🤔

Reply:
Not sure if we should/can account for this edge case. I think this would mean we could never really get rid of it. In any case, I think this method needs some clean-up.

Reply:
@licitdev How do we feel about this now? I think I'd still prefer to completely remove the `getVersionSaves(ById)` methods in favor of the new source of truth `delta` field.

Reply:
@paescuj I believe we can leave this in for the initial release and remove/refactor it in a subsequent/next one. 

The case where not all versions having the delta due to being created during the migration is a valid one as this is a fairly large table for most users and can take time to update.

We can create an issue to track this TODO to ensure we do not forget to remove in a subsequent release.

Reply:
I guess this feels a bit odd to me because up to this point, we've never accounted for such a case, and I general like to avoid such remnants in the code base as well as the added complexity (if it is not quite clear whether it is really necessary).

If we want to do so
- we should definitely comment on the part of the code which is about to be removed, besides tracking it in an issue 
- under that assumption, we also need to make sure we don't delete any of the version records from the revisions collection in the meantime as it could still contain the deltas
- in the future, I feel like we need a better solution for such cases

> Additionally as the changes would now be backwards compatible we can remove the migration from here and add it to that subsequent release refactor. In doing so we reduce the number of migrations needed to run as well as removing this edge case from occurring during migration.

That would however mean, that the migration needs to be re-written in a way where it doesn't fail if the alter statements have already been executed before.

Reply:
> I guess this feels a bit odd to me because up to this point, we've never accounted for such a case, and I generally like to avoid such remnants in the code base as well as the added complexity (if it is not quite clear whether it is really necessary).

We do a similar thing for the sister PR https://github.com/directus/directus/pull/22295 where we leave the API backwards compatible. IMO it is something we should do for these sorts of edge cases as it would be a more seamless upgrade experience.

> If we want to do so
> 
> * we should definitely comment on the part of the code which is about to be removed, besides tracking it in an issue

Agreed
> * under that assumption, we also need to make sure we don't delete any of the version records from the revisions collection in the meantime as it could still contain the deltas

My understanding (@licitdev can confirm) is we do not need the revisions for computing the delta anymore as it is now additive to existing delta as opposed to a re-compute.

> * in the future, I feel like we need a better solution for such cases

Agreed

> > Additionally as the changes would now be backwards compatible we can remove the migration from here and add it to that subsequent release refactor. In doing so we reduce the number of migrations needed to run as well as removing this edge case from occurring during migration.
> 
> That would however mean, that the migration needs to be re-written in a way where it doesn't fail if the alter statements have already been executed before.

There would no longer be a migration for this PR as changes are backwards compatible and therefore no risk of the alter failing due to being executed previously.

Reply:
> We do a similar thing for the sister PR https://github.com/directus/directus/pull/22295 where we leave the API backwards compatible. IMO it is something we should do for these sorts of edge cases as it would provide a better upgrade experience.

Hmm, but in the case of this PR it isn't specifically about the API but more about the compatibility with a previous DB schema (version `delta` in `directus_revisions`).

> My understanding (@licitdev can confirm) is we do not need the revisions for computing the delta anymore as it is now additive to existing delta as opposed to a re-compute.

The assumption stated in https://github.com/directus/directus/pull/22413#discussion_r1595372786, is that versions might be added during the migration via a not-yet-migrated system, meaning it would still be tracked in `directus_revisions` instead of `directus_versions`. This would also mean, we still need to treat `directus_revisions` as a source of truth, thus cannot delete version records in there.
So in a follow-up version, we would need a second migration, which migrates these remaining records. Which also means, for large instances where this edge-case is expected to happen, we would need to ensure that both version updates are not skipped, but performed individually.
(Technically, under such an assumption, we can never be completely sure that all `directus_revisions` records have been migrated, as the same edge-case could occur again during the second migration-phase, though of course much less likely - which to me is a sign that we should definitely find a more suitable solution if we want to cover such cases)

@licitdev Please correct me if I misunderstood something!

Reply:
@paescuj The migration has been simplified in https://github.com/directus/directus/pull/22413/commits/3c3083a12a9728afce7d92de1a020d0d94366a6b so that we will only need to run the data migration once in a subsequent version which also removes the logic for legacy versions.